### PR TITLE
CONSF-38: Create Core class and implement simple observable Model class

### PR DESCRIPTION
**Pull request by kenkouot**
Stubs out `Core` class module and also implements a `Model` class built on top of a micro-EventEmitter implementation.
### Proposed API & Usage

``` es6
const piXmlDocument = /** fetched from mw **/;
const builder = new InfoboxTemplateBuilder({
  from: piXmlDocument
});

// Set some data on the infobox (note, this setter API is not set up for the ordered nature of the infobox data structure yet, so expect additional methods that support the Infobox schema's specifics
builder.data.set('foo', 1);
// Set some theme data (which is a straight flat k:v structure
builder.theme.set('color', '#fff');
// serialize and persist the data to MW
builder.save();

// Usage from UI
// from inside a view
// MyGenericView
{
  init() {
    builder.data.on('propertyDidChange', this.render);
    builder.theme.on('propertyDidChange', this.buildStyles);
  },

  render() {/* does something */},

  buildStyles() {/* does something */} 

  onUserClickedSaveButton() {
    builder.save().then(/* profit */);
  } 
}
```

### CODE REVIEW

File: .jshintrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (kenkouot) [inline .jshintrc:1]: Just to get my linter to stop complaining, will put in the the real properly configured environment files in another PR.

File: src/core.js [hunk 2]
```diff
@@ -1,6 +1,44 @@
 'use strict';
+import {Model} from './model';
+import {serialize, deserialize} from './serializers/xml';
+import {persist} from './adapters/mediawiki';
+
+export {Core as InfoboxTemplateBuilder};
```
Comment (lizlux) [inline src/core.js]: Can you export a class before it's defined?

File: src/core.js [hunk 3]
```diff
@@ -1,6 +1,44 @@
...
+
+		const {from} = params;
+
+		/*
+		 * If builder is instantiated with a serialized document, we will deconstruct it
+		 * into our internal representation, and populate the builder with those values
+		 */
+		if (from) {
+
+			[this.data, this.theme] = deserialize(from);
```
Comment (lizlux) [inline src/core.js]: Looks like we don't get to destructure objects like `{foo: 'foo', bar: 'bar'}` using array syntax, only arrays, like `['foo', 'bar']`. Error I'm seeing is `Uncaught (in promise) Error: Invalid attempt to destructure non-iterable instance`.

**Comment by lizlux** [comment]
What's the plan for adding unit tests?

File: src/core.js (hunk 2 above)
Comment (kenkouot) [inline src/core.js]: Yep, they are hoisted in ES6, and imports/exports at the top seems to be the de facto best practice. Note, class definitions are not hoisted in TypeScript.

Comment (lizlux) [inline src/core.js]: I haven't tested it, I just read https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Hoisting so I thought I'd ask.

Comment (kenkouot) [inline src/core.js]: Hmmm, maybe you're right. I'll move the export statement to the bottom.

**Comment by lizlux** [comment]
Just the one question about unit tests, but code LGTM.

**Comment by lizlux** [comment]
Ah, just saw https://wikia-inc.atlassian.net/browse/CONSF-49, so :+1:

**Pull request merged**